### ISSUE TITLE
[sql] Support `db.operation.batch.size` attribute

### ISSUE THREAD

**Issue by alanwest** [open-telemetry/opentelemetry-dotnet-contrib#2240]
Refer to the [conventions](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/database/mssql.md) for attribute descriptions.

Question: How is batch size determined?

**Comment by martincostello** [comment]
Looks like this might be changed: https://github.com/open-telemetry/semantic-conventions/issues/1984

**Comment by stevejgordon** [comment]
@martincostello Feel free to assign to me. I'm on leave next week though, so can pick it up when I'm back. Unless we need to wait on the open issue you linked.

**Comment by alanwest** [comment]
Removing this from the milestone for the stable release of the SqlClient instrumentation. New attributes can be added to spans after stability. This will not be considered a breaking change.

**Comment by martincostello** [comment]
FYI there were some updates to the spec for this in [v1.43.0](https://github.com/open-telemetry/semantic-conventions/releases/tag/v1.43.0) - I've updated the link in the description to match.

**Comment by PetarJerinic** [comment]
Went through the code, and here are my findings on the questions from the description:

1. `SqlBatch` can't be leveraged from `DiagnosticSource` subscriber since it never reaches subscriber. `SqlBatch.ExecuteNonQuery()` delegates to `SqlCommand` that doesn't have a publicly exposed batch count. There is  ([_RPCList](https://github.com/dotnet/SqlClient/blob/v7.0.2/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlCommand.Batch.cs#L43)) in `SqlCommand.Batch` that keeps count, but is not a public member.
2. AddBatchCommand overwrites CommandText ([SqlCommand.Batch.cs#L25](https://github.com/dotnet/SqlClient/blob/v7.0.2/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlCommand.Batch.cs#L25)), and due to it, there is only the last statement of the batch.
3. I focused on Java, seems ([OpenTelemetryStatement.java](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/v2.30.0/instrumentation/jdbc/library/src/main/java/io/opentelemetry/instrumentation/jdbc/internal/OpenTelemetryStatement.java)) instrumentation wraps `Statement` and increments the counter. Java implementation is a wrapper and can make it work that way.

I would suggest that the maintainers of `dotnet/SqlClient` are asked to expose the batch on the diagnostic payload. Any thoughts? Alternatively, we could read `dotnet/SqlClient` internals, but any future changes there would break the functionality.

**Comment by martincostello** [comment]
Yeah, ideally sqlclient would expose this somehow, then we would consume it. Using internals via reflection can cause issues long-term - there was a similar issue elsewhere recently cause by doing this when StackExchange.Redis v3 was released.

**Comment by PetarJerinic** [comment]
Filed on dotnet/SqlClient#4545
